### PR TITLE
feat(lint): forbid workspace-module mocks behind a shrink-only ledger, add fake S3 for tests

### PR DESCRIPTION
**Pull request by jan-kubica**
## Rule

`no-internal-module-mock` (test scope): `mock.module` against a relative, `@/`, or `@stll/` specifier, or a non-literal one, is an error. npm packages and runtime builtins stay allowed as external boundaries.

A module mock replaces the dependency graph beneath a test with a fabrication, so the test keeps passing after the mocked module's contract changes. Bun's mock registry is also process-wide, which is why the API runner has to batch files around it.

Existing pairs are grandfathered in `scripts/internal-module-mock-ledger.json` (`<file>::<specifier>`, sorted, duplicate-free). The rule reports a listed pair whose mock is gone as stale, so the only sanctioned edit is deleting a line; the ratchet metric `internal-module-mock-ledger-entries` fails when the ledger grows. The rule is tracked in the `test-integrity` suppression tier.

## Test infrastructure

- `tests/helpers/fake-s3.ts`: an in-process S3-compatible store (`Bun.serve`): object GET/HEAD/PUT/DELETE, server-side copy, ListObjectsV2, S3 XML error codes, presigned GETs, per-request failure injection, optional in-flight delay. `configureS3ForTesting` / `configureS3PresignForTesting` point `lib/s3` and `lib/s3-presign` at it (endpoint override only; null in production). A smoke suite drives every real helper through both transports.
- `CONTENT_ENCRYPTION_KEY` is set in the test environment, so the real AES-GCM implementation runs in tests. `content-encryption.ts` gains its first behavioural suite (round trip, cross-organization rejection, tamper detection, fresh IV).

## Migrations (ledger 284 → 244)

CLI env and login tests inject the environment tier and the browser/loopback boundaries; `env.ts` derives its variable names from the documented constants. API tests that mocked `lib/s3`, `lib/s3-presign`, and `content-encryption` now run the real modules and assert on what landed in the store, under which key and content type, and what decrypts.

Each migrated file was mutation-checked: a production change the old mock could not observe (wrong key, wrong bucket, dropped content type, dropped tenant signing scope, plaintext instead of ciphertext, wrong env-var name) left the old test green and fails the migrated one. All mutations were reverted; no production behaviour changes.

Two tests in the corpus-cancellation suite asserted an ordering that existed only in their mock; they now assert the invariant it protected (the store's denial is reported and no sibling object lands).

https://claude.ai/code/session_01AEpRTRcShGxDR7kv99T9VT

## Summary by CodeRabbit

- **New Features**
  - Added validation to prevent tests from mocking internal workspace modules.
  - Added a tracked, shrink-only process for managing existing exceptions.
  - Added configurable test endpoints for object storage and a reusable in-memory storage service.
  - Added support for the documented machine API key environment variable.

- **Tests**
  - Expanded coverage for storage, encryption, uploads, cleanup, copying, authentication and presigned URLs.
  - Replaced fragile module mocks with realistic service simulations and end-to-end assertions.

### CODE REVIEW

File: scripts/ratchet.ts [hunk 1]
```diff
@@ -1222,6 +1224,19 @@ const PER_RULE_SUPPRESSION_METRICS: readonly RatchetMetric[] =
...
+// "<test file>::<workspace module>" pair. The rule
+// (.oxlint-plugins/no-internal-module-mock.ts) reports an unlisted pair and a
+// listed pair whose mock is gone; this budget is what stops a new pair from
+// being listed instead of fixed.
+const countInternalModuleMockLedgerEntries: FileCounter = (content) => {
+  const parsed: unknown = JSON.parse(content);
+  if (!Array.isArray(parsed)) {
+    panic("internal-module-mock ledger must be a JSON array");
+  }
+  return parsed.length;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply scripts/ratchet.ts:1237]: Agreed: the count alone cannot see a swap. c83aa46 adds `scripts/check-internal-module-mock-ledger.ts`, run by `verify.sh` next to the ratchet: every ledger entry must already exist in the base branch's ledger, so the set can only lose members (self-test plus a swap mutation both exercised). The ratchet metric stays as the length cap.

File: .oxlint-plugins/no-internal-module-mock.ts [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+        let bindings: MockBinding[] = [];
+
+        return {
+          before() {
+            ledgerFile = ledgerFileFor(filenameForContext(context));
+            seen = new Set();
+            bindings = [];
+          },
+          ImportDeclaration(node) {
+            bindings.push(...readMockBindings(node));
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply .oxlint-plugins/no-internal-module-mock.ts]: Fixed in 52df6c8: bindings are collected from the Program body before any call is visited, and the fixture now places the first flagged call above its import.

File: apps/api/src/tests/helpers/fake-s3.ts [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+      });
+    }
+
+    const id = objectId(bucket, key);
+    if (method === "COPY" && copySourceKey !== null) {
+      const source = objects.get(objectId(bucket, copySourceKey));
+      if (source === undefined) {
+        return errorResponse("NoSuchKey", 404, copySourceKey);
+      }
+      objects.set(id, source);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/tests/helpers/fake-s3.ts]: Fixed in 52df6c8: put and copy snapshot the bytes (`slice()`), so a later mutation of a caller buffer or a source object cannot rewrite a stored object.

File: scripts/lint-suppressions.ts [hunk 4]
```diff
@@ -195,6 +195,12 @@ export const TRACKED_SUPPRESSION_RULES = [
     guards:
       "throw assertions that pin no error, so any unrelated failure keeps them green",
   },
+  {
+    rule: "no-internal-module-mock/no-internal-module-mock",
+    tier: "test-integrity",
+    guards:
+      "module mocks of workspace modules, which keep a test green after the mocked contract changes",
+  },
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply scripts/lint-suppressions.ts:203]: Fixed in 52df6c8: the rule is in the ratchet fixture budget table (`no-internal-module-mock/no-internal-module-mock: 0`); this was also the cause of the code-quality failure.

File: scripts/ratchet.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jan-kubica) [reply scripts/ratchet.ts:1237]: Addressed in c83aa46, outside the ratchet: `scripts/check-internal-module-mock-ledger.ts` (run by verify.sh next to the ratchet) requires every ledger entry to exist in the base branch's ledger, so a same-count replacement fails; its self-test covers the swap case. Sortedness and uniqueness are enforced by the rule when it loads the ledger. The ratchet metric stays as the length cap only.

File: .ai/local-skills/conventions-testing/SKILL.md [hunk 5]
```diff
@@ -52,6 +52,15 @@ enforcement, branded types) live in
...
   edge-case inputs, or isolating external services
+- Do not `mock.module` a workspace module (`@/...`, `@stll/...`, a
+  relative path). The fabricated module never drifts with the real one, so
+  the test keeps passing after the contract it depends on changes, and Bun's
+  mock registry is process-wide. Inject the collaborator instead (handler
+  context, an options parameter) and pass a plain fake; if the module wraps
+  an external boundary, mock the npm package it calls or expose a test seam.
+  `no-internal-module-mock` enforces this; existing pairs are grandfathered
+  in `scripts/internal-module-mock-ledger.json`, which only shrinks: delete a
+  line when you remove its mock, never add one.
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply .ai/local-skills/conventions-testing/SKILL.md:64]: Applied in faed211 (in the skill source under .ai/local-skills and regenerated), worded as "always name the target with a string literal".
